cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/resource/KafkaBrokerConfigurationDiff.java

```diff
@@ -0,0 +1,205 @@
+/*
+ * Copyright Strimzi authors.
+ * License: Apache License 2.0 (see the file LICENSE or http://apache.org/licenses/LICENSE-2.0.html).
+ */
+
+package io.strimzi.operator.cluster.operator.resource;
+
+import com.fasterxml.jackson.databind.JsonNode;
+import io.fabric8.kubernetes.api.model.ConfigMap;
+import io.fabric8.zjsonpatch.JsonDiff;
+import io.strimzi.operator.cluster.model.KafkaConfiguration;
+import io.strimzi.operator.cluster.model.KafkaVersion;
+import io.strimzi.operator.cluster.model.ModelUtils;
+import org.apache.kafka.clients.admin.AlterConfigOp;
+import org.apache.kafka.clients.admin.Config;
+import org.apache.kafka.clients.admin.ConfigEntry;
+import org.apache.kafka.common.config.ConfigResource;
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.Map;
+import java.util.Optional;
+import java.util.regex.Pattern;
+
+import static io.fabric8.kubernetes.client.internal.PatchUtils.patchMapper;
+
+/**
+ * Computes a diff between current config supplied as a Map ConfigResource, Config and desired config supplied as a ConfigMap
+ * An algorithm:
+ *  1. Create map from supplied desired ConfigMap
+ *  2. Fill placeholders (e.g. ${BROKER_ID}) in desired map
+ *  3. Put all entries from desired to the diff
+ *  4a. If the entry is in IGNORABLE_PROPERTIES or entry.value from desired is equal to entry.value from current, remove entry from diff
+ *  4b. If entry was removed from desired, add it to the diff with default value.
+ *      If custom entry was removed, add it to the diff with 'null' value.
+ *
+ */
+public class KafkaBrokerConfigurationDiff {
+
+    private static final Logger log = LogManager.getLogger(KafkaBrokerConfigurationDiff.class.getName());
+    private final Map<ConfigResource, Config> current;
+    private Collection<ConfigEntry> currentEntries;
+    private ConfigMap desired;
+    private KafkaConfiguration diff;
+    private KafkaVersion kafkaVersion;
+    private int brokerId;
+    private ArrayList<String> deletedEntries;
+    Map<ConfigResource, Collection<AlterConfigOp>> updated = new HashMap<>();
+
+    public static final Pattern IGNORABLE_PROPERTIES = Pattern.compile(
+            "^(broker\\.id"
```

Comment:
Why are these ignored? Because they use placeholders? It would be nice to have some clarification in coments.

Reply:
I am not sure filling up the placeholders is the correct way. The values are filled after the pod is rolled

Reply:
So my question here is really about where did these fields come from and if we can get some explanatory comments.

Reply:
I think the comment here would be still handy. You do not need to explain every field. Just say "these are the fields we always rool for right now because ...